File: CS/MvpvmNavigation/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using MvpvmNavigation.Model;
using DevExpress.Mvvm;
using DevExpress.XtraNavBar;
using DevExpress.Mvvm.POCO;
using MvpvmNavigation.ViewModels;
using MvpvmNavigation.Presenter;
using DevExpress.XtraBars.Navigation;
using DevExpress.XtraBars;
using DevExpress.Utils;
using DevExpress.Utils.Animation;


namespace MvpvmNavigation {
    public partial class Form1 : DevExpress.XtraBars.Ribbon.RibbonForm, ISupportTransitions {
        public Form1() {
            InitializeComponent();
            this.viewModelCore = ViewModelSource.Create<MainFormViewModel>();
            this.presenterCore = CreatePresenter();
            ViewModel.RegisterServices(this);
        }
        MainFormViewModel viewModelCore;
        public MainFormViewModel ViewModel {
            get { return viewModelCore; }
        }
        MainFormPresenter presenterCore;
        public MainFormPresenter Presenter {
            get { return presenterCore; }
        }
        protected virtual MainFormPresenter CreatePresenter() {
            return new MainFormPresenter(panelControl1, ViewModel);
        }
        protected override void OnClosed(EventArgs e) {
            Presenter.Dispose();
            base.OnClosed(e);
        }
        protected override void OnLoad(System.EventArgs e) {
            base.OnLoad(e);
            ModuleType[] modules = new ModuleType[] { 
                ModuleType.ModuleA, ModuleType.ModuleB 
            };
            ViewModel.SelectedModuleType = modules[0];
            RegisterNavigationItems(navBarControl1, modules);
            RegisterNavigationMenuItems(barSubItem1, modules);
        }
        void RegisterNavigationItems(NavBarControl navBar, ModuleType[] modules) {
            NavBarGroup[] groups = new NavBarGroup[modules.Length];
            for(int i = 0; i < modules.Length; i++) {
                ModuleType moduleType = modules[i];
                NavBarGroup navGroup = new NavBarGroup();
                navGroup.Tag = modules[i];
                navGroup.Name = "navGroup" + ViewModel.GetName(moduleType);
                navGroup.Caption = ViewModel.GetCaption(moduleType);
                groups[i] = navGroup;
            }
            navBar.Groups.AddRange(groups);
        }
        void RegisterNavigationMenuItems(BarLinkContainerItem menuItem, ModuleType[] modules) {
            BarItem[] items = new BarItem[modules.Length];
            for(int i = 0; i < modules.Length; i++) {
                ModuleType moduleType = modules[i];
                BarCheckItem biModule = new BarCheckItem();
                biModule.Caption = ViewModel.GetCaption(moduleType);
                biModule.Name = "biModule" + ViewModel.GetName(moduleType);
                biModule.GroupIndex = 1;
                biModule.BindCommand((t) => ViewModel.SelectModule(t), ViewModel, () => moduleType);
                items[i] = biModule;
            }
            menuItem.AddItems(items);
        }
        void officeNavigationBar1_RegisterItem(object sender, NavigationBarNavigationClientItemEventArgs e) {
            NavBarGroup navGroup = (NavBarGroup)e.NavigationItem;
            ModuleType moduleType = (ModuleType)navGroup.Tag;
            e.Item.Tag = moduleType;
            e.Item.Text = ViewModel.GetCaption(moduleType);
            e.Item.Name = "navItem" + ViewModel.GetName(moduleType);
            e.Item.BindCommand((t) => ViewModel.SelectModule(t), ViewModel, () => moduleType);
        }

        #region ISupportTransitions Members
        void ISupportTransitions.StartTransition(bool forward, object waitParameter) {
            var transition = transitionManager1.Transitions[panelControl1];
            var animator = transition.TransitionType as DevExpress.Utils.Animation.SlideFadeTransition;
            animator.Parameters.EffectOptions = forward ? PushEffectOptions.FromRight : PushEffectOptions.FromLeft;
            if (waitParameter == null)
                transition.ShowWaitingIndicator = DefaultBoolean.False;
            else {
                transition.ShowWaitingIndicator = DefaultBoolean.True;
                transition.WaitingIndicatorProperties.Caption = ViewModel.GetCaption((ModuleType)waitParameter);
                transition.WaitingIndicatorProperties.Description = "Loading...";
                transition.WaitingIndicatorProperties.ContentMinSize = new System.Drawing.Size(160, 0);
            }
            transitionManager1.StartTransition(panelControl1);
        }
        void ISupportTransitions.EndTransition() {
            transitionManager1.EndTransition();
        }
        #endregion
    }
}

File: VB/MvpvmNavigation/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports MvpvmNavigation.Model
Imports DevExpress.Mvvm
Imports DevExpress.XtraNavBar
Imports DevExpress.Mvvm.POCO
Imports MvpvmNavigation.ViewModels
Imports MvpvmNavigation.Presenter
Imports DevExpress.XtraBars.Navigation
Imports DevExpress.XtraBars
Imports DevExpress.Utils
Imports DevExpress.Utils.Animation


Namespace MvpvmNavigation
    Partial Public Class Form1
        Inherits DevExpress.XtraBars.Ribbon.RibbonForm
        Implements ISupportTransitions

        Public Sub New()
            InitializeComponent()
            Me.viewModelCore = ViewModelSource.Create(Of MainFormViewModel)()
            Me.presenterCore = CreatePresenter()
            ViewModel.RegisterServices(Me)
        End Sub
        Private viewModelCore As MainFormViewModel
        Public ReadOnly Property ViewModel() As MainFormViewModel
            Get
                Return viewModelCore
            End Get
        End Property
        Private presenterCore As MainFormPresenter
        Public ReadOnly Property Presenter() As MainFormPresenter
            Get
                Return presenterCore
            End Get
        End Property
        Protected Overridable Function CreatePresenter() As MainFormPresenter
            Return New MainFormPresenter(panelControl1, ViewModel)
        End Function
        Protected Overrides Sub OnClosed(ByVal e As EventArgs)
            Presenter.Dispose()
            MyBase.OnClosed(e)
        End Sub
        Protected Overrides Sub OnLoad(ByVal e As System.EventArgs)
            MyBase.OnLoad(e)
            Dim modules() As ModuleType = { ModuleType.ModuleA, ModuleType.ModuleB }
            ViewModel.SelectedModuleType = modules(0)
            RegisterNavigationItems(navBarControl1, modules)
            RegisterNavigationMenuItems(barSubItem1, modules)
        End Sub
        Private Sub RegisterNavigationItems(ByVal navBar As NavBarControl, ByVal modules() As ModuleType)
            Dim groups(modules.Length - 1) As NavBarGroup
            For i As Integer = 0 To modules.Length - 1
                Dim moduleType As ModuleType = modules(i)
                Dim navGroup As New NavBarGroup()
                navGroup.Tag = modules(i)
                navGroup.Name = "navGroup" & ViewModel.GetName(moduleType)
                navGroup.Caption = ViewModel.GetCaption(moduleType)
                groups(i) = navGroup
            Next i
            navBar.Groups.AddRange(groups)
        End Sub
        Private Sub RegisterNavigationMenuItems(ByVal menuItem As BarLinkContainerItem, ByVal modules() As ModuleType)
            Dim items(modules.Length - 1) As BarItem
            For i As Integer = 0 To modules.Length - 1
                Dim moduleType As ModuleType = modules(i)
                Dim biModule As New BarCheckItem()
                biModule.Caption = ViewModel.GetCaption(moduleType)
                biModule.Name = "biModule" & ViewModel.GetName(moduleType)
                biModule.GroupIndex = 1
                biModule.BindCommand(Sub(t) ViewModel.SelectModule(t), ViewModel, Function() moduleType)
                items(i) = biModule
            Next i
            menuItem.AddItems(items)
        End Sub
        Private Sub officeNavigationBar1_RegisterItem(ByVal sender As Object, ByVal e As NavigationBarNavigationClientItemEventArgs) Handles officeNavigationBar1.RegisterItem
            Dim navGroup As NavBarGroup = CType(e.NavigationItem, NavBarGroup)
            Dim moduleType As ModuleType = DirectCast(navGroup.Tag, ModuleType)
            e.Item.Tag = moduleType
            e.Item.Text = ViewModel.GetCaption(moduleType)
            e.Item.Name = "navItem" & ViewModel.GetName(moduleType)
            e.Item.BindCommand(Sub(t) ViewModel.SelectModule(t), ViewModel, Function() moduleType)
        End Sub

        #Region "ISupportTransitions Members"
        Private Sub ISupportTransitions_StartTransition(ByVal forward As Boolean, ByVal waitParameter As Object) Implements ISupportTransitions.StartTransition
            Dim transition = transitionManager1.Transitions(panelControl1)
            Dim animator = TryCast(transition.TransitionType, DevExpress.Utils.Animation.SlideFadeTransition)
            animator.Parameters.EffectOptions = If(forward, PushEffectOptions.FromRight, PushEffectOptions.FromLeft)
            If waitParameter Is Nothing Then
                transition.ShowWaitingIndicator = DefaultBoolean.False
            Else
                transition.ShowWaitingIndicator = DefaultBoolean.True
                transition.WaitingIndicatorProperties.Caption = ViewModel.GetCaption(DirectCast(waitParameter, ModuleType))
                transition.WaitingIndicatorProperties.Description = "Loading..."
                transition.WaitingIndicatorProperties.ContentMinSize = New System.Drawing.Size(160, 0)
            End If
            transitionManager1.StartTransition(panelControl1)
        End Sub
        Private Sub ISupportTransitions_EndTransition() Implements ISupportTransitions.EndTransition
            transitionManager1.EndTransition()
        End Sub
        #End Region
    End Class
End Namespace
